src/disk_io_thread.cpp

```diff
@@ -196,16 +197,23 @@ namespace libtorrent
 
 	void disk_io_thread::abort(bool wait)
 	{
-		m_abort = true;
-		if (num_threads() == 0)
+		// abuse the job mutex to make setting m_abort and checking the thread count atomic
+		// see also the comment in thread_fun
+		std::unique_lock<std::mutex> l(m_job_mutex);
+		if (m_abort.exchange(true)) return;
```

Comment:
There are a couple of m_abort checks down in the job handlers which don't need to hold the job mutex. Now that I've looked at it though, m_num_running_threads does not need to be atomic anymore.